lib/worker.js

```diff
@@ -0,0 +1,210 @@
+'use strict';
+
+if (!process.features.experimental_workers) {
+  throw new Error('Experimental workers are disabled');
+}
+
+const util = require('util');
+const assert = require('assert');
+const EventEmitter = require('events');
+const WorkerContextBinding = process.binding('WorkerContext');
+const JSONStringify = function(value) {
+  if (value === undefined) value = null;
+  return JSON.stringify(value);
+};
+const JSONParse = JSON.parse;
+const EMPTY_ARRAY = [];
+
+const workerContextSymbol = Symbol('workerContext');
+const installEventsSymbol = Symbol('installEvents');
+const checkAliveSymbol = Symbol('checkAlive');
+const initSymbol = WorkerContextBinding.initSymbol;
+
+const builtinErrorTypes = new Map([
+  Error, SyntaxError, RangeError, URIError, TypeError, EvalError, ReferenceError
+].map(function(Type) {
+  return [Type.name, Type];
+}));
+
+const Worker = WorkerContextBinding.JsConstructor;
+util.inherits(Worker, EventEmitter);
+
+Worker.prototype[initSymbol] = function(entryModulePath, options) {
+  if (typeof entryModulePath !== 'string')
+    throw new TypeError('entryModulePath must be a string');
+  EventEmitter.call(this);
+  options = Object(options);
+  var keepAlive = options.keepAlive === undefined ? true : !!options.keepAlive;
+  this[workerContextSymbol] =
+      new WorkerContextBinding.WorkerContext(entryModulePath,
+                                             {keepAlive: keepAlive});
+  this[installEventsSymbol]();
+};
+
+Worker.prototype[installEventsSymbol] = function() {
+  const workerObject = this;
+  const workerContext = this[workerContextSymbol];
+
+  const onerror = function(payload) {
+    var ErrorConstructor = builtinErrorTypes.get(payload.builtinType);
+    if (typeof ErrorConstructor !== 'function')
+      ErrorConstructor = Error;
+    const error = new ErrorConstructor(payload.message);
+    error.stack = payload.stack;
+    util._extend(error, payload.additionalProperties);
+    workerObject.emit('error', error);
+  };
+
+  workerContext._onexit = function(exitCode) {
+    workerObject[workerContextSymbol] = null;
+    workerObject.emit('exit', exitCode);
+  };
+
+  workerContext._onmessage = function(payload, messageType) {
+    payload = JSONParse(payload);
+    switch (messageType) {
+      case WorkerContextBinding.USER:
+        return workerObject.emit('message', payload);
+      case WorkerContextBinding.INTERNAL:
+        assert.fail('unreachable');
+      case WorkerContextBinding.EXCEPTION:
+        return onerror(payload);
+      default:
+        assert.fail('unreachable');
+    }
```

Comment:
It looks like `this[workerContext]` is a new object instance. So you could do like `this[workerContextSymbol].parent = this;`, then from `_onmessage` do `onerror.call(this.parent, payload)` then in `onerror()` just use `this.emit('error', error);`. Also `_onexit` could then just use `this.parent[workerConetxtSymbol] = null;` and such.

Reason for this is two fold. First to allow hoisting all the functions. Second is just from personal experience that debugging is easier when you can transverse the object to get all the information you need instead of looking at the function execution context for variable values.

Whether you decide to or not, I don't consider this a blocker for the PR.